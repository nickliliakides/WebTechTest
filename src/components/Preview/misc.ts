import { createStyles, Theme } from '@material-ui/core/styles';

export const styles = (theme: Theme) =>
  createStyles({
    root: {
      margin: 0,
      padding: theme.spacing(2),
    },
    closeButton: {
      position: 'absolute',
      right: theme.spacing(1),
      top: theme.spacing(1),
      color: theme.palette.grey[500],
    },
  });

export const shuffle = (array: string[]): string[] => {
  let currentIndex = array.length,
    temporaryValue: string,
    randomIndex: number;

  // While there remain elements to shuffle...
  while (0 !== currentIndex) {
    // Pick a remaining element...
    randomIndex = Math.floor(Math.random() * currentIndex);
    currentIndex -= 1;

    // And swap it with the current element.
    temporaryValue = array[currentIndex];
    array[currentIndex] = array[randomIndex];
    array[randomIndex] = temporaryValue;
  }

  return array;
};

export const mockQuestion = {
  id: 1,
  category: 'test',
  type: 'test',
  difficulty: 'test',
  question: 'test',
  correct_answer: 'test',
  incorrect_answers: ['test', 'test', 'test'],
};
