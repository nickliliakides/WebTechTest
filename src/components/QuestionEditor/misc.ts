import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';

// Main Styling
export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      '& .MuiTextField-root': {
        margin: theme.spacing(1),
        width: '40vw',
        textAlign: 'left',
        fontFamily: 'Roboto, sans-serif',
      },
    },
    main_heading: {
      textAlign: 'center',
    },
    headings: {
      padding: '8px',
    },
    paper: {
      marginRight: 20,
      padding: 20,
      opacity: '.85',
      backgroundColor: '#e6e6e6',
    },
    paper_inner: {
      margin: 8,
      padding: 20,
      backgroundColor: 'white',
      border: '2px groove #0e4b16',
      boxShadow: '0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23)',
    },
    flex: {
      display: 'flex',
    },
    left: {
      flex: 3,
    },
    right: {
      flex: 1,
      display: 'flex',
      justifyContent: 'center',
    },
    btn_container: {
      display: 'flex',
      justifyContent: 'space-between',
    },
    save_btn: {
      backgroundColor: 'green',
      margin: 10,
      '&:hover': {
        backgroundColor: 'green',
        opacity: 0.8,
      },
    },
    add_btn: {
      backgroundColor: 'orange',
      margin: 10,
      '&:hover': {
        backgroundColor: 'orange',
        opacity: 0.8,
      },
    },
    preview_btn: {
      backgroundColor: 'blue',
      margin: 10,
      '&:hover': {
        backgroundColor: 'blue',
        opacity: 0.8,
      },
    },
    delete_icon: {
      color: 'red',
      position: 'relative',
      right: '38px',
      top: '22px',
      marginRight: '-25px',
      cursor: 'pointer',
    },
  })
);

// Initial Component's State
export const initialState = {
  formdata: {
    question: '',
    answer1: '',
    answer2: '',
    answer3: '',
    answer4: '',
    answer5: '',
    checked1: true,
    checked2: false,
    checked3: false,
    checked4: false,
    checked5: false,
    closed3: true,
    closed4: true,
    closed5: true,
  },
  submitted: false,
};
