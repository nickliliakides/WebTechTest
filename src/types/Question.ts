export interface QuestionResponse {
  response_code: number;
  results?: Question[] | null;
}

export interface Question {
  id: number;
  category: string;
  type: string;
  difficulty: string;
  question: string;
  correct_answer: string;
  incorrect_answers: string[];
}
