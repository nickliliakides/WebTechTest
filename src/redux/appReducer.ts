import { createStore, combineReducers, applyMiddleware } from 'redux';

import createSagaMiddleware from 'redux-saga';

import rootSaga from './sagas/rootSaga';

import * as sportwebservice from './actions/questionwebservice';

import { Question } from '../types/Question';

export interface QuizState {
  questions: Question[];
  selectedQuestion: Question | {};
}

const initialQuizState: QuizState = {
  questions: [],
  selectedQuestion: {},
};

function questionReduxReducer(
  state = initialQuizState,
  action: { type: string; payload: any }
) {
  const { type, payload } = action;

  switch (type) {
    case sportwebservice.Types.GET_QUESTIONS_SUCCESS:
      return {
        ...state,
        questions: [...payload.results, ...state.questions],
      };
    case sportwebservice.Types.DELETE_QUESTION_REQUEST:
      return {
        ...state,
        questions: [
          ...state.questions.filter((question) => question.id !== payload),
        ],
      };
    case sportwebservice.Types.SET_SELECTED_QUESTION:
      return {
        ...state,
        selectedQuestion: payload,
      };
    case sportwebservice.Types.UPDATE_SELECTED_QUESTION:
      return {
        questions: [
          payload,
          ...state.questions.filter((question) => question.id !== payload.id),
        ],
        selectedQuestion: {},
      };
    default:
      return state;
  }
}

const overallReducer = combineReducers({
  quiz: questionReduxReducer,
});

// Sage Middleware is used for fetching external resources.
const sagaMiddleware = createSagaMiddleware();

const store = createStore(overallReducer, applyMiddleware(sagaMiddleware));

sagaMiddleware.run(rootSaga);

export { store };
