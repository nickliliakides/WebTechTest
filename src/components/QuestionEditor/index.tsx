import React, { useEffect, useRef, useState } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { updateSelectedQuestion } from '../../redux/actions/questionwebservice';
import { toast } from 'react-toastify';
import Fade from 'react-reveal/Fade';
import { withStyles } from '@material-ui/core/styles';
import { green } from '@material-ui/core/colors';
import { FormControlLabel, Grid, Paper } from '@material-ui/core';
import { Button } from '@material-ui/core';
import SaveIcon from '@material-ui/icons/Save';
import VisibilityIcon from '@material-ui/icons/Visibility';
import DeleteForeverIcon from '@material-ui/icons/DeleteForever';
import AddBoxIcon from '@material-ui/icons/AddBox';
import Checkbox, { CheckboxProps } from '@material-ui/core/Checkbox';
import { ValidatorForm, TextValidator } from 'react-material-ui-form-validator';
import CircularProgress from '@material-ui/core/CircularProgress';
import { Question } from '../../types/Question';
import Preview from '../Preview';
import { useStyles, initialState } from './misc';

const GreenCheckbox = withStyles({
  root: {
    color: green[400],
    '&$checked': {
      color: green[600],
    },
  },
  checked: {},
})((props: CheckboxProps) => <Checkbox color='default' {...props} />);

interface IProps {
  dispatch: Dispatch;
  selectedQuestion: Question;
}

const QuestionEditor: React.FC<IProps> = ({
  dispatch,
  selectedQuestion,
}): JSX.Element => {
  const classes = useStyles();
  const [state, setState] = useState(initialState);
  const [preview, setPreview] = useState(false);
  const refForm = useRef('form');

  useEffect(() => {
    if (selectedQuestion && selectedQuestion.id) {
      const { question, correct_answer, incorrect_answers } = selectedQuestion;
      const formdata = { ...state.formdata };
      formdata['question'] = question;
      formdata['answer1'] = correct_answer;
      formdata['answer2'] = incorrect_answers[0];
      formdata['answer3'] = incorrect_answers[1] ? incorrect_answers[1] : '';
      formdata['answer4'] = incorrect_answers[2] ? incorrect_answers[2] : '';
      formdata['answer5'] = incorrect_answers[3] ? incorrect_answers[3] : '';
      formdata['closed3'] = incorrect_answers[1] ? false : true;
      formdata['closed4'] = incorrect_answers[2] ? false : true;
      formdata['closed5'] = incorrect_answers[3] ? false : true;
      setState({
        ...state,
        formdata,
      });
    }
  }, [selectedQuestion]); // eslint-disable-line react-hooks/exhaustive-deps

  const {
    question,
    answer1,
    answer2,
    answer3,
    answer4,
    answer5,
    checked1,
    checked2,
    checked3,
    checked4,
    checked5,
    closed3,
    closed4,
    closed5,
  } = state.formdata;

  const handleCheckBoxChange = (
    event: React.ChangeEvent<HTMLInputElement>
  ): void => {
    const formdata = { ...state.formdata };
    formdata[event.target.name] = event.target.checked;
    setState({ ...state, formdata });
  };

  const handleTextFieldChange = (
    event: React.ChangeEvent<HTMLInputElement>
  ): void => {
    const formdata = { ...state.formdata };
    formdata[event.target.name] = event.target.value;
    setState({
      ...state,
      formdata,
    });
  };

  const handleDelete = (field: string, fieldState: string): void => {
    const formdata = { ...state.formdata };
    formdata[field] = '';
    formdata[fieldState] = true;
    setState({
      ...state,
      formdata,
    });
  };

  const handleNewInput = (): void => {
    const displayInput = (field: string): void => {
      const formdata = { ...state.formdata };
      formdata[field] = false;
      setState({ ...state, formdata });
    };

    if (closed3) {
      displayInput('closed3');
    }

    if (closed4) {
      displayInput('closed4');
    }

    if (closed5) {
      displayInput('closed5');
    }
  };

  const handleSubmit = (): void => {
    if (!selectedQuestion.id) {
      toast.error(
        'There is not selected question to edit. Please select one from the list on the right.'
      );
      return;
    }

    const correctQuestionsToCheck = [
      checked1,
      checked2,
      checked3,
      checked4,
      checked5,
    ];
    let correctLength = 0;
    correctQuestionsToCheck.forEach((q) => {
      if (q === true) correctLength++;
    });

    let fieldsAreValid = question !== '' && answer1 !== '' && answer2 !== '';

    if (!fieldsAreValid) return;

    if (fieldsAreValid && correctLength !== 1) {
      toast.error('There must be only 1 correct answer!');
      return;
    }

    const correct_answer = checked1
      ? answer1
      : checked2
      ? answer2
      : checked3
      ? answer3
      : checked4
      ? answer4
      : answer5;

    const incorrectAnswersToCheck = checked1
      ? [answer2, answer3, answer4, answer5]
      : checked2
      ? [answer1, answer3, answer4, answer5]
      : checked3
      ? [answer1, answer2, answer4, answer5]
      : checked4
      ? [answer1, answer2, answer3, answer5]
      : [answer1, answer2, answer3, answer4];

    const filterIncorrectAnswers = (): string[] => {
      const result: string[] = [];
      incorrectAnswersToCheck.forEach((ans) => {
        if (ans !== '') {
          result.push(ans);
        }
      });
      return result;
    };

    const dataToSubmit = {
      ...selectedQuestion,
      question,
      correct_answer,
      incorrect_answers: filterIncorrectAnswers(),
    };
    dispatch(updateSelectedQuestion(dataToSubmit));
    setState({ ...state, submitted: true });
    setTimeout(() => {
      setState(initialState);
    }, 800);
  };

  return (
    <Grid data-testid='editor' className='edit_container' item={true} md={7}>
      <div className='editor'>
        <h1 className={classes.main_heading}>Edit Question</h1>
        <ValidatorForm
          className={classes.root}
          ref={refForm}
          onSubmit={handleSubmit}
        >
          <Fade top duration={700} delay={500}>
            <Paper className={classes.paper}>
              <div className={classes.paper_inner}>
                <h2 data-testid='header' className={classes.headings}>
                  Question
                </h2>
                <TextValidator
                  id='question'
                  data-testid='question-field'
                  name='question'
                  label='Question'
                  value={question}
                  variant='outlined'
                  onChange={handleTextFieldChange}
                  validators={['required']}
                  errorMessages={['This field is required']}
                  disabled={!selectedQuestion.id ? true : false}
                />
                <div className={classes.flex}>
                  <h2 className={`${classes.headings} ${classes.left}`}>
                    Answers
                  </h2>
                  <h2 className={`${classes.headings} ${classes.right}`}>
                    Correct
                  </h2>
                </div>
                <div className={classes.flex}>
                  <TextValidator
                    className={classes.left}
                    name='answer1'
                    label='Answer'
                    value={answer1}
                    variant='outlined'
                    onChange={handleTextFieldChange}
                    validators={['required']}
                    errorMessages={['This field is required']}
                    disabled={!selectedQuestion.id ? true : false}
                  />
                  <FormControlLabel
                    className={classes.right}
                    control={
                      <GreenCheckbox
                        checked={checked1}
                        onChange={handleCheckBoxChange}
                        name='checked1'
                        data-testid='checkbox'
                      />
                    }
                    label=''
                    disabled={!selectedQuestion.id ? true : false}
                  />
                </div>
                <div className={classes.flex}>
                  <TextValidator
                    className={classes.left}
                    name='answer2'
                    label='Answer'
                    value={answer2}
                    variant='outlined'
                    onChange={handleTextFieldChange}
                    validators={['required']}
                    errorMessages={['This field is required']}
                    disabled={!selectedQuestion.id ? true : false}
                  />

                  <FormControlLabel
                    className={classes.right}
                    control={
                      <GreenCheckbox
                        checked={checked2}
                        onChange={handleCheckBoxChange}
                        name='checked2'
                      />
                    }
                    label=''
                    disabled={!selectedQuestion.id ? true : false}
                  />
                </div>
                {(answer3 !== '' || !closed3) && (
                  <div className={classes.flex}>
                    <TextValidator
                      className={classes.left}
                      name='answer3'
                      label='Answer'
                      value={answer3}
                      variant='outlined'
                      onChange={handleTextFieldChange}
                      validators={['required']}
                      errorMessages={['This field is required']}
                      disabled={answer3 === '' && closed3 ? true : false}
                    />
                    <DeleteForeverIcon
                      onClick={() => handleDelete('answer3', 'closed3')}
                      className={classes.delete_icon}
                      name={answer3}
                      data-testid='delete-icon'
                    />
                    <FormControlLabel
                      className={classes.right}
                      control={
                        <GreenCheckbox
                          checked={checked3}
                          onChange={handleCheckBoxChange}
                          name='checked3'
                        />
                      }
                      label=''
                      disabled={closed3}
                    />
                  </div>
                )}
                {(answer4 !== '' || !closed4) && (
                  <div className={classes.flex}>
                    <TextValidator
                      className={classes.left}
                      name='answer4'
                      label='Answer'
                      value={answer4}
                      variant='outlined'
                      onChange={handleTextFieldChange}
                      validators={['required']}
                      errorMessages={['This field is required']}
                      disabled={answer4 === '' && closed4 ? true : false}
                    />
                    <DeleteForeverIcon
                      onClick={() => handleDelete('answer4', 'closed4')}
                      className={classes.delete_icon}
                    />
                    <FormControlLabel
                      className={classes.right}
                      control={
                        <GreenCheckbox
                          checked={checked4}
                          onChange={handleCheckBoxChange}
                          name='checked4'
                        />
                      }
                      label=''
                      disabled={closed4}
                    />
                  </div>
                )}
                {(answer5 !== '' || !closed5) && (
                  <div className={classes.flex}>
                    <TextValidator
                      className={classes.left}
                      name='answer5'
                      label='Answer'
                      value={answer5}
                      variant='outlined'
                      onChange={handleTextFieldChange}
                      validators={['required']}
                      errorMessages={['This field is required']}
                      disabled={answer5 === '' && closed5 ? true : false}
                    />
                    <DeleteForeverIcon
                      onClick={() => handleDelete('answer5', 'closed5')}
                      className={classes.delete_icon}
                    />
                    <FormControlLabel
                      className={classes.right}
                      control={
                        <GreenCheckbox
                          checked={checked5}
                          onChange={handleCheckBoxChange}
                          name='checked5'
                        />
                      }
                      label=''
                      disabled={closed5}
                    />
                  </div>
                )}
                <div>
                  <Button
                    data-testid='save-button'
                    className={classes.save_btn}
                    variant='contained'
                    color='primary'
                    type='submit'
                    startIcon={
                      state.submitted ? (
                        <CircularProgress size={20} />
                      ) : (
                        <SaveIcon />
                      )
                    }
                    disabled={state.submitted || !selectedQuestion.id}
                  >
                    Save
                  </Button>
                  <Button
                    data-testid='add-button'
                    className={classes.add_btn}
                    variant='contained'
                    color='primary'
                    type='button'
                    disabled={
                      (!closed3 && !closed4 && !closed5) || !selectedQuestion.id
                    }
                    startIcon={<AddBoxIcon />}
                    onClick={handleNewInput}
                  >
                    Add Answer
                  </Button>
                  <Button
                    data-testid='preview-button'
                    className={classes.preview_btn}
                    variant='contained'
                    color='primary'
                    type='button'
                    disabled={!selectedQuestion.id}
                    startIcon={<VisibilityIcon />}
                    onClick={() => {
                      if (selectedQuestion.id) {
                        setPreview(true);
                        setTimeout(() => setPreview(false), 200);
                      }
                    }}
                  >
                    Preview
                  </Button>
                </div>
              </div>
            </Paper>
          </Fade>
        </ValidatorForm>
        <Preview show={preview} question={selectedQuestion} />
      </div>
    </Grid>
  );
};

const mapStateToProps = (state: { quiz: { selectedQuestion: Question } }) => {
  return { selectedQuestion: state.quiz.selectedQuestion };
};

export default connect(mapStateToProps)(QuestionEditor);
