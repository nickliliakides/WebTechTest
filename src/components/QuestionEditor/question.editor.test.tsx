import React from 'react';
import ReactDOM from 'react-dom';
import QuestionEditor from './index';
import { Provider } from 'react-redux';
import { Dispatch, Store } from 'redux';
import configureStore from 'redux-mock-store';
import '@testing-library/jest-dom/extend-expect';
import { Question } from '../../types/Question';
import renderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import { mockQuestion } from '../Preview/misc';

const mockStore = configureStore([]);

interface IProps {
  dispatch: Dispatch;
  selectedQuestion: Question;
}

describe('Question editor component tests', () => {
  let store: Store;
  let mockDispatch: jest.Mock;
  let mockProps: IProps;
  let element: JSX.Element;

  beforeEach(() => {
    store = mockStore({
      quiz: {
        selectedQuestion: mockQuestion,
      },
    });
    mockDispatch = jest.fn();

    mockProps = {
      dispatch: mockDispatch,
      selectedQuestion: mockQuestion,
    };

    element = (
      <Provider store={store}>
        <QuestionEditor {...mockProps} />
      </Provider>
    );
  });

  it('Matches saved snapshot', () => {
    const tree = renderer.create(element).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(element, div);
  });

  it('Question header has text content "Question"', () => {
    const { getByTestId } = render(element);
    expect(getByTestId('header')).toHaveTextContent('Question');
  });

  it('Question field is empty when component is mounted', () => {
    const { getByTestId } = render(element);
    expect(getByTestId('question-field').querySelector('input')).toBeEmpty();
    setTimeout(() => {
      expect(
        getByTestId('question-field').querySelector('input')
      ).toBeDisabled();
    }, 400);
  });

  it('Checkbox is checked after component mounted', () => {
    const { getByTestId } = render(element);
    expect(getByTestId('checkbox').querySelector('input')).toBeChecked();
    setTimeout(() => {
      expect(getByTestId('checkbox').querySelector('input')).toBeDisabled();
    }, 400);
  });

  it('Answer delete icon is in the document', () => {
    const { getByTestId } = render(element);
    expect(getByTestId('delete-icon')).toBeInTheDocument();
  });

  it('Save button is visible and disabled after component mounted ', () => {
    const { getByTestId } = render(element);
    const button = getByTestId('save-button');
    expect(button).toBeVisible();
    setTimeout(() => {
      expect(button).toBeDisabled();
    }, 400);
  });

  it('Add answer button is visible and disabled after component mounted ', () => {
    const { getByTestId } = render(element);
    const button = getByTestId('add-button');
    expect(button).toBeVisible();
    setTimeout(() => {
      expect(button).toBeDisabled();
    }, 400);
  });

  it('Preview button is visible and disabled after component mounted ', () => {
    const { getByTestId } = render(element);
    const button = getByTestId('preview-button');
    expect(button).toBeVisible();
    setTimeout(() => {
      expect(button).toBeDisabled();
    }, 400);
  });
});
