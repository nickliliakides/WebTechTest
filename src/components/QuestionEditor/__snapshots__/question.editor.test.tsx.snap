// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question editor component tests Matches saved snapshot 1`] = `
<div
  className="MuiGrid-root edit_container MuiGrid-item MuiGrid-grid-md-7"
  data-testid="editor"
>
  <div
    className="editor"
  >
    <h1
      className="makeStyles-main_heading-2"
    >
      Edit Question
    </h1>
    <form
      className="makeStyles-root-1"
      onSubmit={[Function]}
    >
      <div
        className="react-reveal"
        style={
          Object {
            "opacity": undefined,
          }
        }
      >
        <div
          className="MuiPaper-root makeStyles-paper-4 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <h2
            className="makeStyles-headings-3"
            data-testid="header"
          >
            Question
          </h2>
          <div>
            <div
              className="MuiFormControl-root MuiTextField-root"
              data-testid="question-field"
            >
              <label
                className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink={false}
                htmlFor="question"
                id="question-label"
              >
                Question
              </label>
              <div
                className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  autoFocus={false}
                  className="MuiInputBase-input MuiOutlinedInput-input"
                  disabled={false}
                  id="question"
                  name="question"
                  onAnimationStart={[Function]}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden={true}
                  className="PrivateNotchedOutline-root-13 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    className="PrivateNotchedOutline-legendLabelled-15"
                  >
                    <span>
                      Question
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            className="makeStyles-flex-5"
          >
            <h2
              className="makeStyles-headings-3 makeStyles-left-6"
            >
              Answers
            </h2>
            <h2
              className="makeStyles-headings-3 makeStyles-right-7"
            >
              Correct
            </h2>
          </div>
          <div
            className="makeStyles-flex-5"
          >
            <div>
              <div
                className="MuiFormControl-root MuiTextField-root makeStyles-left-6"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                  data-shrink={false}
                >
                  Answer
                </label>
                <div
                  className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiInputBase-input MuiOutlinedInput-input"
                    disabled={false}
                    name="answer1"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden={true}
                    className="PrivateNotchedOutline-root-13 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      className="PrivateNotchedOutline-legendLabelled-15"
                    >
                      <span>
                        Answer
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <label
              className="MuiFormControlLabel-root makeStyles-right-7"
            >
              <span
                aria-disabled={false}
                className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-19 MuiCheckbox-root Component-root-17 PrivateSwitchBase-checked-20 Mui-checked Component-checked-18"
                data-testid="checkbox"
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={null}
              >
                <span
                  className="MuiIconButton-label"
                >
                  <input
                    checked={true}
                    className="PrivateSwitchBase-input-22"
                    data-indeterminate={false}
                    disabled={false}
                    name="checked1"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                    />
                  </svg>
                </span>
              </span>
              <span
                className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
              >
                
              </span>
            </label>
          </div>
          <div
            className="makeStyles-flex-5"
          >
            <div>
              <div
                className="MuiFormControl-root MuiTextField-root makeStyles-left-6"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                  data-shrink={false}
                >
                  Answer
                </label>
                <div
                  className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiInputBase-input MuiOutlinedInput-input"
                    disabled={false}
                    name="answer2"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden={true}
                    className="PrivateNotchedOutline-root-13 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      className="PrivateNotchedOutline-legendLabelled-15"
                    >
                      <span>
                        Answer
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <label
              className="MuiFormControlLabel-root makeStyles-right-7"
            >
              <span
                aria-disabled={false}
                className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-19 MuiCheckbox-root Component-root-17"
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={null}
              >
                <span
                  className="MuiIconButton-label"
                >
                  <input
                    checked={false}
                    className="PrivateSwitchBase-input-22"
                    data-indeterminate={false}
                    disabled={false}
                    name="checked2"
                    onChange={[Function]}
                    type="checkbox"
                  />
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                </span>
              </span>
              <span
                className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
              >
                
              </span>
            </label>
          </div>
          <div>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-save_btn-9 MuiButton-containedPrimary"
              data-testid="save-button"
              disabled={false}
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="submit"
            >
              <span
                className="MuiButton-label"
              >
                <span
                  className="MuiButton-startIcon MuiButton-iconSizeMedium"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
                    />
                  </svg>
                </span>
                Save
              </span>
            </button>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-add_btn-10 MuiButton-containedPrimary"
              data-testid="add-button"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                <span
                  className="MuiButton-startIcon MuiButton-iconSizeMedium"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
                    />
                  </svg>
                </span>
                Add Answer
              </span>
            </button>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-preview_btn-11 MuiButton-containedPrimary"
              data-testid="preview-button"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                <span
                  className="MuiButton-startIcon MuiButton-iconSizeMedium"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                    />
                  </svg>
                </span>
                Preview
              </span>
            </button>
          </div>
        </div>
      </div>
    </form>
    <div />
  </div>
</div>
`;
